### DISCUSSION TITLE
Requirements for BinaryType template parameter?

### DISCUSSION THREAD

**Discussion by epvbergen** [nlohmann/json#2511] [Q&A]
Hi, thank you for an excellent piece of infrastructure.

I'm trying to use a custom BinaryType with basic_json. However I've been unsucessful in passing anything other than std::vector<uint8_t> there. Even a plain std::vector<intptr_t> fails to compile with nlohmann::json 3.8.0 and GCC 9.3.0. 

Can you say anything more regarding the requirements for BinaryType? I understand it needs to be a sequence container as you want to provide the option of dumping as an array instead of specialized serialization. But what about the container's value type? What restrictions are there by design and what restrictions are there by accident?

Second question: if a conversion from binary to JSON array wouldn't have been required, which I understand is only a debugging feature, couldn't the type requirements be relaxed so that the BinaryType doesn't need to implement a full STL container interface? 

Background:

One of my biggest uses for nlohmann::json is its capability to easily store and pass arbitrarily nested trees of named variant values, to slice them and access their values in a type safe manner. An universal, structured, dynamically typed value, like a Python dict or a Javascript object. Serialization and parsing are only secondary and only used whenever the data leaves the process.

Even if supplying a recursive variant datastructure with a very user friendly and STL-compatible API is not the intended use for this library, it does this very well. Duplicating all that in something home grown because "nlohmann::json is intended for parsing and serialization of JSON and JSON-like data structures" seems like a waste of effort, especially considering how much of nlohmann::json in terms of LOC isn't concerned with serialization or parsing but with representing the parsed data and providing a nice interface for that.

From this perspective it's becoming a burden that basic_json insists it owns its data and cannot own references that point inside buffers owned elsewhere, neither the buffers they were parsed from, nor user supplied data. I've been looking at ways to parameterize one of the value types, especially the recently introduced BinaryType, to see if I can make that happen.

But as said, I can't seem to put anything other than std::vector<uint8_t> there. Even other integral types in the vector don't work.

As far as implementation is concerned, could it be an idea to move responsibility for (de)serialization and dumping to the binary subtype wrapper (binary_t = nlohmann::byte_container_with_subtype<BinaryType>) that was recently added, and using that wrapper as the BinaryType parameter? 

A default BinaryWrapper which users can use as a base class could provide the serializations for the existing serializations and subtypes.

This change would allow zero-copy parsing without SAX and could provide a way for the user to implement even more data types of the various JSON supersets to be implemented, without putting that burden on the library itself.

**Comment by nlohmann** [comment]
BinaryType has no formally documented concept, but from the implementation it needs to be default-constructible, copy/move-constructible, and support push_back/.data()/.size(), because byte_container_with_subtype<BinaryType> derives directly from it (include/nlohmann/byte_container_with_subtype.hpp). More importantly, the binary serializers in include/nlohmann/detail/output/binary_writer.hpp assume the container's value_type is exactly one byte wide: e.g. write_characters() does reinterpret_cast<const CharType*>(j.m_data.m_value.binary->data()) and writes binary->size() raw bytes from that pointer, which is only correct for byte-sized elements. That's why std::vector<uint8_t> works while std::vector<intptr_t> does not. As of the current version, the zero-copy / arbitrary-element-type redesign you proposed (moving (de)serialization responsibility onto byte_container_with_subtype itself so non-byte containers could be used) has not been implemented — BinaryType still needs to behave like a byte container today.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
